File: server/internal/sync/merge_test.go
package sync

import (
	"sort"
	"testing"
)

// These tests mirror app/src/lib/local/merge.test.ts — the client and server
// merge logic must stay behaviorally identical.

const (
	t1 = "000000000000001:00000:devA"
	t2 = "000000000000002:00000:devB"
	t3 = "000000000000003:00000:devA"
)

func TestFieldClockAcceptNewerWins(t *testing.T) {
	fc := ParseFieldClock("{}")
	if !fc.Accept("name", t2) {
		t.Fatal("newer stamp must be accepted")
	}
	if fc["name"] != t2 {
		t.Fatalf("clock not recorded: %v", fc)
	}
	if fc.Accept("name", t1) {
		t.Fatal("older stamp must be rejected")
	}
	if fc.Accept("name", t2) {
		t.Fatal("equal stamp must be rejected")
	}
}

func TestFieldClockFieldsIndependent(t *testing.T) {
	fc := FieldClock{"name": t3}
	if !fc.Accept("note", t1) {
		t.Fatal("an older stamp on a different field must be accepted")
	}
}

func TestFieldClockTolerantParsing(t *testing.T) {
	if got := ParseFieldClock("not json"); len(got) != 0 {
		t.Fatalf("malformed clock: %v", got)
	}
	if got := ParseFieldClock(""); len(got) != 0 {
		t.Fatalf("empty clock: %v", got)
	}
}

func TestORSetAddRemove(t *testing.T) {
	s := ORSet{}
	s.Add("p1", t1)
	if got := s.Values(); len(got) != 1 || got[0] != "p1" {
		t.Fatalf("after add: %v", got)
	}
	s.Remove("p1")
	if got := s.Values(); len(got) != 0 {
		t.Fatalf("after remove: %v", got)
	}
	s.Add("p1", t3) // re-add with a new tag becomes visible again
	if got := s.Values(); len(got) != 1 {
		t.Fatalf("after re-add: %v", got)
	}
}

func TestORSetAddWinsAgainstConcurrentRemove(t *testing.T) {
	// Device A removes what it observed; device B's concurrent add carries an
	// unobserved tag. After merge the element must stay visible.
	a := ORSet{}
	a.Add("p1", t1)
	a.Remove("p1")

	b := ORSet{}
	b.Add("p1", t2)

	a.Merge(b)
	if got := a.Values(); len(got) != 1 || got[0] != "p1" {
		t.Fatalf("add must win: %v", got)
	}
}

func TestORSetRemoveTagsKeepsUnlistedTags(t *testing.T) {
	// The server row has two adds; the delta says the origin observed only t1.
	s := ORSet{}
	s.Add("p1", t1)
	s.Add("p1", t2)
	s.RemoveTags("p1", []string{t1})
	if got := s.Values(); len(got) != 1 {
		t.Fatalf("tag t2 was not observed by the remover — element must stay: %v", got)
	}
	s.RemoveTags("p1", []string{t2})
	if got := s.Values(); len(got) != 0 {
		t.Fatalf("all tags tombstoned — element must vanish: %v", got)
	}
}

func TestORSetMergeCommutative(t *testing.T) {
	mk := func() ORSet {
		s := ORSet{}
		s.Add("x", t1)
		s.Remove("x")
		return s
	}
	other := ORSet{}
	other.Add("x", t2)
	other.Add("y", t3)

	ab := mk()
	ab.Merge(other)

	ba := ORSet{}
	ba.Merge(other)
	ba.Merge(mk())

	va, vb := ab.Values(), ba.Values()
	sort.Strings(va)
	sort.Strings(vb)
	if len(va) != len(vb) {
		t.Fatalf("orders diverge: %v vs %v", va, vb)
	}
	for i := range va {
		if va[i] != vb[i] {
			t.Fatalf("orders diverge: %v vs %v", va, vb)
		}
	}
}

func TestORSetTolerantParsing(t *testing.T) {
	if got := ParseORSet("nope"); len(got) != 0 {
		t.Fatalf("malformed set: %v", got)
	}
	if got := ParseORSet("null"); len(got) != 0 {
		t.Fatalf("null set: %v", got)
	}
}

File: app/src/lib/local/merge.test.ts
import { describe, it, expect } from 'vitest';
import {
  parseFieldClock, accept, parseORSet, orAdd, orRemove, orMerge, orValues,
  type ORSet
} from './merge';

const t1 = '000000000000001:00000:devA';
const t2 = '000000000000002:00000:devB';
const t3 = '000000000000003:00000:devA';

describe('per-field LWW (FieldClock)', () => {
  it('accepts a newer stamp and records it', () => {
    const fc = parseFieldClock('{}');
    expect(accept(fc, 'name', t2)).toBe(true);
    expect(fc.name).toBe(t2);
  });

  it('rejects an older or equal stamp', () => {
    const fc = { name: t2 };
    expect(accept(fc, 'name', t1)).toBe(false);
    expect(accept(fc, 'name', t2)).toBe(false);
    expect(fc.name).toBe(t2); // unchanged
  });

  it('fields are independent', () => {
    const fc = { name: t3 };
    expect(accept(fc, 'note', t1)).toBe(true); // other field, older stamp: fine
  });

  it('tolerates malformed clock JSON', () => {
    expect(parseFieldClock('not json')).toEqual({});
    expect(parseFieldClock(null)).toEqual({});
    expect(parseFieldClock(undefined)).toEqual({});
  });

  it('same millisecond: counter then node id break the tie deterministically', () => {
    const a = '000000000000005:00001:devA';
    const b = '000000000000005:00002:devA';
    const fc = { f: a };
    expect(accept(fc, 'f', b)).toBe(true);
  });
});

describe('OR-Set (add-wins)', () => {
  it('add makes the element visible', () => {
    const s: ORSet = {};
    orAdd(s, 'p1', t1);
    expect(orValues(s)).toEqual(['p1']);
  });

  it('remove hides only observed adds', () => {
    const s: ORSet = {};
    orAdd(s, 'p1', t1);
    orRemove(s, 'p1');
    expect(orValues(s)).toEqual([]);
  });

  it('concurrent add wins over remove (add-wins)', () => {
    // Device A removes based on what it saw; device B adds concurrently with
    // a tag A has not observed. After merge, the element stays.
    const a: ORSet = {};
    orAdd(a, 'p1', t1);
    orRemove(a, 'p1');

    const b: ORSet = {};
    orAdd(b, 'p1', t2); // unobserved by A's remove

    orMerge(a, b);
    expect(orValues(a)).toEqual(['p1']);
  });

  it('merge is commutative: both orders converge', () => {
    const mk = () => {
      const s: ORSet = {};
      orAdd(s, 'x', t1);
      orRemove(s, 'x');
      return s;
    };
    const other: ORSet = {};
    orAdd(other, 'x', t2);
    orAdd(other, 'y', t3);

    const ab = mk(); orMerge(ab, other);
    const ba: ORSet = JSON.parse(JSON.stringify(other)); orMerge(ba, mk());
    expect(orValues(ab).sort()).toEqual(orValues(ba).sort());
  });

  it('re-add after remove is visible again', () => {
    const s: ORSet = {};
    orAdd(s, 'p1', t1);
    orRemove(s, 'p1');
    orAdd(s, 'p1', t3);
    expect(orValues(s)).toEqual(['p1']);
  });

  it('duplicate tags are not accumulated', () => {
    const s: ORSet = {};
    orAdd(s, 'p1', t1);
    orAdd(s, 'p1', t1);
    expect(s['p1'].a).toEqual([t1]);
  });

  it('tolerates malformed set JSON', () => {
    expect(parseORSet('nope')).toEqual({});
    expect(parseORSet(null)).toEqual({});
    expect(parseORSet('null')).toEqual({});
  });
});
